File: algorithms-robert-sedgewick/sort/QuickSort.cpp
#include "QuickSort.h"

void QuickSort::sort()
{
    sort(0, dataLen - 1);
}

void QuickSort::sort(int lo, int hi)
{
    if (lo >= hi)
    {
        return;
    }

    int p = partition(lo, hi);
    sort(lo, p - 1);
    sort(p + 1, hi);
}

int QuickSort::partition(int lo, int hi)
{
    int v = data[lo];
    int i = lo + 1;
    int j = hi;

    while (true)
    {
        while (less(data[i], v))
        {
            if (i == hi)
                break;
            i++;
        }

        while (less(v, data[j]))
        {
            if (j == lo)
                break;
            j--;
        }

        if (i >= j)
            break;

        exch(i, j);
    }

    exch(lo, j);

    return j; // data[lo..j-1] <= data[j] <= data[j+1..hi]
}

File: leetcode-cpp/java_solutions/src/algo/basic/QuickSort.java
package algo.basic;

/**
 * 超级简洁的快速排序算法，我第一次看到的时候惊呆了
 * Created by clpsz on 2016/9/14.
 */
public class QuickSort {
    public void sort(int[] nums) {
        sort(nums, 0, nums.length-1);
    }

    private void sort(int[] nums, int lo, int hi) {
        if (lo >= hi) return;
        int p = partition(nums, lo, hi);
        sort(nums, lo, p-1);
        sort(nums, p+1, hi);
    }

    private int partition(int[] nums, int lo, int hi) {
        if (lo >= hi) return -1;

        int i = lo+1;
        int j = lo+1;

        while (j <= hi) {
            if (nums[j] < nums[lo]) {
                swap(nums, j, i);
                i++;
            }

            j++;
        }
        swap(nums, lo, i-1);

        return i-1;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public static void main(String[] args) {
        int[] given = new int[]{3,2,1,5,6,4};

        QuickSort quickSort = new QuickSort();
        quickSort.sort(given);

        for (int n : given) {
            System.out.println(n);
        }
    }
}
